### PR TITLE
[Enhancement] Increase programming exercise test stability

### PR DESCRIPTION
**Pull request by jhainzlschmid**
### Motivation and Context

In the ProgrammingExerciseParticipation tests cypress interacts with the online code editor to type in submissions. Since cypress actually types the content of a file like a user would the code editor applies autocompletion like automatically inserting a closing bracket. So if cypress types:
```java
private void foo() {
}
```
the content in the online code editor will be:
```java
private void foo() {
}
}
```
This results in a build error. As a workaround we currently tell cypress to type shift + pagedown + del to essentially delete all brackets the code editor added on it's own. Sometimes cypress seems to fail to enter that key combination which results in no deletion of the unwanted content and as a consequence the test fails because of a build error. The error occured for example in [this build](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-119/log) and also occured when @lschlesinger tested this [cypress PR](https://github.com/ls1intum/Artemis/pull/3829).

### Description

Instead of using the somewhat unreliable keycombination shift + pagedown + del we go for the brute force way of simply send 100 keystrokes of the del key. This also removes the unwanted content, but is more reliable. Furthermore it is also faster because we don't have to force a delay between the keystrokes.
I also added an assertion. which makes sure that the deleteFile query actually returns HTTP 200, because I saw that fail sometimes on the test server. In that case the test failure is now more understandable.

### Steps for Testing

I executed this PR on the testing branch and the last 5 [builds passed](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-124)

### CODE REVIEW

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts [hunk 1]
```diff
@@ -4,6 +4,15 @@ const buildingAndTesting = 'Building and testing...';
...
 export class OnlineEditorPage {
+    deleteRemainingContent: string;
+
+    constructor() {
+        this.deleteRemainingContent = '';
+        for (let i = 0; i < 100; i++) {
+            this.deleteRemainingContent += '{del}';
+        }
+    }
+
```
Comment (Strohgelaender) [inline src/test/cypress/support/pageobjects/OnlineEditorPage.ts:15]:
```suggestion
    readonly deleteRemainingContent = '{del}'.repeat(100);
```

Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts]: Damn that's a lot better. Thanks! 😄

**Review by FrankeLukas — APPROVED** [review]
Makes sense. It would be better if we somehow could get the file content length and delete using that instead of hardcoding 100, but I don't know if that is possible.

**Comment by jhainzlschmid** [comment]
> Makes sense. It would be better if we somehow could get the file content length and delete using that instead of hardcoding 100, but I don't know if that is possible.

Yeah I agree. I actually tried counting the amount of opening brackets and then simply sending that amount of {del} commands + 20 (buffer to make sure that everything is still deleted if cypress drops some keystrokes), but this didn't work because one {del} does not equal one superfluous closing bracket. The reason for that is that the code editor not only adds the closing bracket, but also adds indentation. So depending on the position of the opening bracket cypress has to send multiple {del} commands (1 to delete the newline + x for the added indentation + 1 to remove the closing bracket). We could count the amount of closing brackets, multiply that by for example 10 and then it would probably work for most implemented classes, but of course there would always be a content, which would be large enough to exceed enough indentation that the factor 10 wouldn't be enough anymore.
A somewhat safe option could be to send as many {del} commands as the content has characters. I will implement and test that.

**Comment by jhainzlschmid** [comment]
I implemented the dynamic deletion. And the [tests](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-135) are still passing

**Review by FrankeLukas — DISMISSED** [review]

**Review by lschlesinger — APPROVED** [review]
I ran the `ProgrammingExercise.spec.ts` two times "locally" with TS2 as base URL, worked fine. Your approach of deleting the automatically added signs is quite creative, it's crazy with which problems you have to deal with when writing such E2E tests 😄

**Review by joschahenningsen — APPROVED** [review]
Worked for me as well

**Review by lschlesinger — CHANGES_REQUESTED** [review]
-> Edit, also run `ProgrammingExerciseParticipation.spec.ts` two times, the result was the following:
1. Run
Bildschirmfoto 2021-08-03 um 08 22 39
2. Run
Bildschirmfoto 2021-08-03 um 08 41 58

**Comment by jhainzlschmid** [comment]
> -> Edit, also run `ProgrammingExerciseParticipation.spec.ts` two times, the result was the following:
> 
> 1. Run
> 
> Bildschirmfoto 2021-08-03 um 08 22 39
> 
> 2. Run
> Bildschirmfoto 2021-08-03 um 08 41 58

Hey, could you change these two fields in the `cypress.json` file and rerun the tests:
```json
{
   "video": true,
   "defaultCommandTimeout": 20000
}
```
The bamboo buildplan also runs with 20s as default timeout. The default timeout will be increased to 20s once this [PR](https://github.com/ls1intum/Artemis/pull/3829) is merged to develop. If the tests still fail you will find a recording of the test under `src/test/cypress/videos`. Could you send that to me then, so I can have a look at what is going on?

**Comment by jhainzlschmid** [comment]
I added a small sleep between typing in the submission and clicking the submit button. A real user would also not immediately click on submit. This seems to fix the failures on ts2. I also reran the [buildplan](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-138) and the tests still pass

**Review by joschahenningsen — APPROVED** [review]
re-approve :)

**Review by lschlesinger — APPROVED** [review]
Run the test 2 times, no problems occured. I changed the .json file as described.

**Pull request merged**